src/libraries/System.Private.Uri/src/System/Uri.cs

```diff
@@ -4650,77 +4594,85 @@ private unsafe char[] GetCanonicalPath(char[] dest, ref int pos, UriFormat forma
             //
             // (path is already  >= 3 chars if recognized as a DOS-like)
             //
-            if (dosPathIdx != 0 && dest[dosPathIdx + pos - 1] == '|')
-                dest[dosPathIdx + pos - 1] = ':';
+            int offset = start + dosPathIdx;
+            if (dosPathIdx != 0 && dest[offset - 1] == '|')
+                dest[offset - 1] = ':';
 
-            if (InFact(Flags.ShouldBeCompressed))
+            if (InFact(Flags.ShouldBeCompressed) && dest.Length - offset > 0)
             {
                 // It will also convert back slashes if needed
-                dest = Compress(dest, (ushort)(pos + dosPathIdx), ref end, _syntax);
-                if (dest[pos] == '\\')
-                    dest[pos] = '/';
+                dest.Length = offset + Compress(dest.RawChars.Slice(offset, dest.Length - offset), _syntax);
+                if (dest[start] == '\\')
+                    dest[start] = '/';
 
                 // Escape path if requested and found as not fully escaped
                 if (formatAs == UriFormat.UriEscaped && NotAny(Flags.UserEscaped) && InFact(Flags.E_PathNotCanonical))
                 {
                     //Note: Flags.UserEscaped check is solely based on trusting the user
-                    dest = UriHelper.EscapeString(new string(dest, pos, end - pos), dest, ref pos, checkExistingEscaped: !IsImplicitFile, '?', '#');
-                    end = pos;
+
+                    // We can't do an in-place escape, create a copy
+                    ValueStringBuilder copy = new ValueStringBuilder(stackalloc char[StackallocThreshold]);
+                    copy.Append(dest.AsSpan(start, dest.Length - start));
+
+                    dest.Length = start;
+                    ReadOnlySpan<char> copySpan = MemoryMarshal.CreateReadOnlySpan(ref copy.GetPinnableReference(), copy.Length);
+                    UriHelper.EscapeString(copySpan, ref dest, checkExistingEscaped: !IsImplicitFile, '?', '#');
+                    start = dest.Length;
+
+                    copy.Dispose();
                 }
             }
             else if (_syntax.InFact(UriSyntaxFlags.ConvertPathSlashes) && InFact(Flags.BackslashInPath))
             {
-                for (int i = pos; i < end; ++i)
+                Span<char> chars = dest.RawChars.Slice(start, dest.Length - start);
+                for (int i = 0; i < chars.Length; i++)
```

Comment:
I don't understand this change.  Why are we slicing to get chars and iterating over its length, but then using `dest`?  Assuming that's a bug and `dest` should be `chars`, seems like we then have a test hole?

Reply:
Right, this should be chars. I will look into getting a test for this

Reply:
This block is currently unreachable as we always add the `ShouldBeCompressed` flag when we have `ConvertPathSlashes`. That looks like I bug - I'll look into cases where the output is changed because of this.